platform/orion/ons.go

```diff
@@ -58,6 +78,59 @@ func (n *NetworkService) Finality() *Finality {
 	return &Finality{finality: n.ons.Finality()}
 }
 
+type NetworkServiceProvider struct {
+	sp              view2.ServiceProvider
+	mutex           sync.RWMutex
+	networkServices map[string]*NetworkService
+}
+
+func NewNetworkServiceProvider(sp view2.ServiceProvider) *NetworkServiceProvider {
+	return &NetworkServiceProvider{sp: sp, networkServices: make(map[string]*NetworkService)}
+}
+
+func (nsp *NetworkServiceProvider) NetworkService(id string) (*NetworkService, error) {
+	nsp.mutex.RLock()
+	ns, ok := nsp.networkServices[id]
+	nsp.mutex.RUnlock()
+	if ok {
+		return ns, nil
+	}
+
+	nsp.mutex.Lock()
+	defer nsp.mutex.Unlock()
+	ns, ok = nsp.networkServices[id]
+	if ok {
+		return ns, nil
+	}
+
```

Comment:
because multiple threads might be waiting at the second lock, the first that get the lock creates the network service and it might be that another thread is looking for the same service. So no need to create it again.